src/atomEffect2.ts

```diff
@@ -0,0 +1,68 @@
+import type { Getter, Setter, WritableAtom } from 'jotai/vanilla'
+import { atom } from 'jotai/vanilla'
+
+type CleanupFn = () => PromiseOrValue<void>
+type PromiseOrValue<T> = Promise<T> | T
+
+export function atomEffect(
+  effectFn: (get: Getter, set: Setter) => PromiseOrValue<void | CleanupFn>
+) {
+  const refAtom = atom<{
+    mounted: boolean
+    rerun: boolean
+    inProgress: number
+    cleanup: CleanupFn | void
+  }>({
+    mounted: false,
+    rerun: false,
+    inProgress: 0,
+    cleanup: undefined,
+  })
+  const initAtom = atom(null, (get, _set, mounted: boolean) => {
+    const ref = get(refAtom)
+    if (mounted) {
+      ref.mounted = true
+    } else {
+      ref.cleanup?.() // do not await
+      ref.mounted = false
+      ref.rerun = false
+      ref.cleanup = undefined
+    }
+  })
+  initAtom.onMount = (init) => {
+    init(true)
+    return () => init(false)
+  }
+
+  const effectAtom = atom(
+    async (get, { setSelf }) => {
+      const ref = get(refAtom)
+      if (!ref.mounted) {
```

Comment:
because ref.mounted is directly written outside of Jotai reactivity, effectAtom will not rerun on mount.

Reply:
Yeah, good point.

On the other hand, do we not need to run effect, if the atom isn't mounted, right?